parlai/core/worlds.py

```diff
@@ -1245,15 +1253,20 @@ def parley(self):
 
         # great, this batch is good to go! let's run it!
         acts = self.world.get_model_agent().batch_act([self._obs[i] for i in batch])
+        self.acts = [[self._task_acts[i] for i in batch], acts]
         # broadcast the results back to all the models
         for i, act in zip(batch, acts):
             # we need to make sure that the teachers saw the result
             self.worlds[i].get_task_agent().observe(act)
             # and that the agent copies saw their own voice
             self.worlds[i].get_model_agent().self_observe(act)
-
             # move these worlds forward
             act = self.worlds[i].get_task_agent().act()
+            # we log the task act and the index of the act
+            # in the buffer for world logging purposes
+            self._task_acts[i] = act
```

Comment:
sorry, could you explain again why we do this twice in `parley`?

Reply:
ah yes so for the indices we selected, we have to move the subworlds forward and get the next task act from the teacher

Reply:
yeah it's dynamic batching. Dynb doesn't in parley() ends with all subworlds ready for the model's turn.